au/code/au/quantity_test.cc

```diff
@@ -717,6 +722,32 @@ TEST(Quantity, MixedTypeSubtractionUsesCommonRepType) {
     EXPECT_THAT(feet(2.f) - feet(1.5), QuantityEquivalent(feet(0.5)));
 }
 
+TEST(Quantity, CommonUnitAlwaysCompletelyIndependentOfOrder) {
+    auto check_units = [](auto unit_a, auto unit_b, auto unit_c) {
+        const auto a = unit_a(1LL);
+        const auto b = unit_b(1LL);
+        const auto c = unit_c(1LL);
+        auto stream_to_string = [](auto x) {
+            std::ostringstream oss;
+            oss << x;
+            return oss.str();
+        };
+        std::vector<std::string> results = {
+            stream_to_string(a + b + c),
+            stream_to_string(a + c + b),
+            stream_to_string(b + a + c),
+            stream_to_string(b + c + a),
+            stream_to_string(c + a + b),
+            stream_to_string(c + b + a),
+        };
+        EXPECT_THAT(results, Each(Eq(results[0])))
+            << "Inconsistency found for (" << a << ", " << b << ", " << c << ")";
+    };
+
+    check_units(centi(meters), miles, meters);
+    check_units(kilo(meters), miles, milli(meters));
+}
```

Comment:
Note to reviewer: if we added only this test case on main, without the fixes, here is what the failure would look like:

```
[ RUN      ] Quantity.CommonUnitAlwaysCompletelyIndependentOfOrder
au/code/au/quantity_test.cc:743: Failure
Value of: results
Expected: only contains elements that is equal to "805177 COM[cm, m, mi]"
  Actual: { "805177 COM[cm, m, mi]", "805177 COM[cm, mi]", "805177 COM[cm, m, mi]", "805177 COM[cm, m, mi]", "805177 COM[cm, mi]", "805177 COM[cm, m, mi]" }, whose element #1 doesn't match
Inconsistency found for (1 cm, 1 mi, 1 m)
```

Reply:
Instead of checking that they are all consistent with the first element, can we check that they are all equal to a specific unit? That might help test readibility.

Reply:
I see the appeal of this suggestion.  I struggled with this for a little bit, but I couldn't see my way to an implementation that did this while keeping the test implementation clean.

One way would be to compare to a desired unit label.  But the unit label for common units will change dramatically after 0.4.0 (#105), and I'd hate to have to update this test for that change.

We could also do it by providing the expected _type_ of the common unit.  But `CommonUnit<...>` isn't supposed to be named directly by end users: we're supposed to use `CommonUnitT<...>`, which performs the de-duping and simplification that is the very thing under test!

Ultimately, I think the current approach is the best option available.  Phrasing it in terms of _order independence_, as mentioned in https://github.com/aurora-opensource/au/issues/295#issuecomment-2374768684, is the most elegant way I know to express what we're really after here.  The downside, as you point out, is that a test failure can't tell you _which one is correct_; all it can do is point out that they should all be the same, and they're not.  But I think in practice this will be good enough to point maintainers in the right direction.